lib/pharos/ssh/client.rb

```diff
@@ -48,23 +59,24 @@ def bastion
       # @param options [Hash] see Net::SSH#start
       def connect(**options)
         synchronize do
-          logger.debug { "connect: #{@user}@#{@host} (#{@opts})" }
-          if bastion
-            gw_opts = {}
-            gw_opts[:keys] = [bastion.ssh_key_path] if bastion.ssh_key_path
-            gateway = Net::SSH::Gateway.new(bastion.address, bastion.user, gw_opts)
-            @session = gateway.ssh(@host, @user, @opts.merge(options))
-          else
-            @session = Net::SSH.start(@host, @user, @opts.merge(options))
+          Pharos::Retry.perform(10, exceptions: CONNECTION_RETRY_ERRORS) do
+            logger.debug { "connect: #{@user}@#{@host} (#{@opts})" }
+            @session = if bastion
+                         bastion.host.ssh.gateway.ssh(@host, @user, @opts.merge(options))
+                       else
+                         Net::SSH.start(@host, @user, @opts.merge(options))
+                       end
           end
         end
       end
 
-      # @param host [String]
-      # @param port [Integer]
-      # @return [Integer] local port number
-      def gateway(host, port)
-        Net::SSH::Gateway.new(@host, @user, @opts).open(host, port)
+      # @return [Net::SSH::Gateway]
+      def gateway
+        @gateway ||= Net::SSH::Gateway.new(@host, @user, @opts).tap do |gw|
+          gw.instance_exec do
+            @thread.report_on_exception = false
```

Comment:
This was added because when rebooting a node that acts as a bastion, the gateway thread for the host with bastion will die from IOError and report the exception to the terminal.

It seems to manage to reconnect with all the connection/reconnection additions in this PR and the exception appears to be harmless.